Support node namespaces

wjwwood opened the pull request:
This is the first change (as it includes the actual storage for the namespace) in a series of several changes up and down the stack. The purpose of these pull requests are to add the notion of a namespace for each node. This namespace will later be used to expand relative topic names and remapping rules.

This pr builds on https://github.com/ros2/rmw/pull/93 and include the commits from there. It should be rebased after that is merged.

This particular pr is relatively simple as it only adds storage for the namespace and adds it to the API for `rmw_create_node()`, see https://github.com/ros2/rmw/commit/14cd7272427b528d5f5ce9e71ea21ab16a3767a2.

I'm going to leave this "in progress" until I get the other pull requests opened.

wjwwood commented:
Ok, now that I got all the pr's open, and took a second look at it, I think these are all ready for review and to be merged after #93.

dirk-thomas commented on rmw/include/rmw/rmw.h:
```diff
@@ -88,7 +88,7 @@ rmw_init(void);
 RMW_WARN_UNUSED
 rmw_node_t *
-rmw_create_node(const char * name, size_t domain_id);
+rmw_create_node(const char * name, const char * name_space, size_t domain_id);
```
Shouldn't this be one word: `namespace`?

wjwwood replied:
It's a keyword in c++ so when this header is included in cpp files it still fails to compile, even though this is inside an extern c block.

dirk-thomas replied:
That was an obvious one :face_with_head_bandage:

I am not sure I like the idea of "separating" the variable name like this though. Some style guide recommend appending an underscore to names which collide with keywords (e.g. PEP 8). Even though this has the potential to collide with the C++ guide for class members I would prefer `namespace_` over `name_space` in this case.

wjwwood replied:
I agree, I'll go through and update all of the prs soon.

dirk-thomas commented:
Can you please also create a PR for `rmw_opensplice`.

wjwwood commented:
Why do I need to open a pr for opensplice?

dirk-thomas commented:
> Why do I need to open a pr for opensplice?

Because we want to continue to be able to build with `rmw_opensplice` in order to test it in the future. Especially if the changes are as minimal as in this case.

wjwwood commented:
I opened https://github.com/ros2/rmw_opensplice/pull/164 and made this commit https://github.com/ros2/rmw_connext/pull/220/commits/2e655933d8b2478a0daffd465f3b052448536888 in order to move this all along, but honestly I think it's waste of time.

I think we should rethink our policy of having opensplice and connext dynamic try to compile on a fresh clone, by disabling them in upstream with an AMENT_IGNORE, add a note to their README's telling people how to re-enable them, and have the farm remove those files when the options are selected.

wjwwood commented:
I rebased this after #93 was merged.

wjwwood commented:
Ok, with two new pr's to fix linger issues I didn't see on my mac, here is CI post rebase:

- Linux:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=2393)](http://ci.ros2.org/job/ci_linux/2393/)
- macOS:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=1843)](http://ci.ros2.org/job/ci_osx/1843/)
- Windows:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=2507)](http://ci.ros2.org/job/ci_windows/2507/)

Still looking for reviews for some of these pr's. Thanks!

wjwwood commented:
@Karsten1987 can you review this pr too?

wjwwood commented:
New CI after addressing comments:

- Linux:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=2397)](http://ci.ros2.org/job/ci_linux/2397/)
- macOS:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=1851)](http://ci.ros2.org/job/ci_osx/1851/)
- Windows:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=2516)](http://ci.ros2.org/job/ci_windows/2516/) (the test failures appear to be unrelated/flaky)

dirk-thomas commented on rmw/include/rmw/validate_node_name.h:
```diff
@@ -37,7 +37,7 @@ extern "C"
  * - must not be an empty string
  * - must only contain alphanumeric characters and underscores (a-z|A-Z|0-9|_)
- * - must not start with an number
+ * - must start with an alphabetical character
```
Maybe I missed the discussion in another thread. Why do we want to prohibit node name starting with an underscore?

wjwwood replied:
I think I was just trying to simplify the rules, but I forgot about the underscores. The implementation didn't actually change. I'll switch it back b6f79db.

wjwwood commented:
In 383db0f I added `rmw_validate_namespace()` as requested here: https://github.com/ros2/rcl/pull/112#discussion_r109942323

In 383db0f I made `invalid_index` optional as requested here: https://github.com/ros2/rcl/pull/112#discussion_r109941394

wjwwood commented:
I believe all the comments on all pull requests have been addressed, I'm looking for final reviews. I'll probably merge later tonight if no one says anything.

Here's the latest CI:

- Linux:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=2409)](http://ci.ros2.org/job/ci_linux/2409/)
- Linux-aarch64:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=45)](http://ci.ros2.org/job/ci_linux-aarch64/45/)
- macOS:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=1858)](http://ci.ros2.org/job/ci_osx/1858/)
- Windows:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=2528)](http://ci.ros2.org/job/ci_windows/2528/) (flaky python service requester test failure only)

dirk-thomas commented on rmw/src/validate_namespace.c:
```diff
@@ -0,0 +1,124 @@
+        break;
+      default:
+        RMW_SET_ERROR_MSG(
+          "rmw_validate_namespace(): unknown rmw_validate_topic_name() validation result"
```
It would be good to include the value of `t_validation_result` in the error message.

wjwwood replied:
I think this should work, it does on macOS at least. I'll need to run CI again to check for warning on Windows: 833c760

wjwwood commented:
Ok, I addressed all the comments and here is passing CI:

- Linux:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=2419)](http://ci.ros2.org/job/ci_linux/2419/)
- Linux-aarch64:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=52)](http://ci.ros2.org/job/ci_linux-aarch64/52/) (existing test failures)
- macOS:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=1872)](http://ci.ros2.org/job/ci_osx/1872/)
- Windows:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=2545)](http://ci.ros2.org/job/ci_windows/2545/) (failing tests are flaky)

wjwwood commented:
I'm going to go ahead and merge, if there are more comments I'll address them afterwards.
